tianshou/policy/modelfree/c51.py

```diff
@@ -0,0 +1,225 @@
+import torch
+import numpy as np
+from numba import njit
+from typing import Any, Dict, Union, Optional, Tuple
+
+from tianshou.policy import DQNPolicy
+from tianshou.data import Batch, ReplayBuffer, to_torch_as, to_numpy
+
+
+class C51Policy(DQNPolicy):
+    """Implementation of Categorical Deep Q-network. arXiv:1707.06887.
+
+    :param torch.nn.Module model: a model following the rules in
+        :class:`~tianshou.policy.BasePolicy`. (s -> logits)
+    :param torch.optim.Optimizer optim: a torch.optim for optimizing the model.
+    :param float discount_factor: in [0, 1].
+    :param int num_atoms: the number of atoms in the support set of the
+        value distribution, defaults to 51.
+    :param float v_min: the value of the smallest atom in the support set,
+        defaults to -10.0.
+    :param float v_max: the value of the largest atom in the support set,
+        defaults to -10.0.
+    :param int estimation_step: greater than 1, the number of steps to look
+        ahead.
+    :param int target_update_freq: the target network update frequency (0 if
+        you do not use the target network).
+    :param bool reward_normalization: normalize the reward to Normal(0, 1),
+        defaults to False.
+
+    .. seealso::
+
+        Please refer to :class:`~tianshou.policy.DQNPolicy` for more detailed
+         explanation.
+    """
+
+    def __init__(
+        self,
+        model: torch.nn.Module,
+        optim: torch.optim.Optimizer,
+        discount_factor: float = 0.99,
+        num_atoms: int = 51,
+        v_min: float = -10.0,
+        v_max: float = 10.0,
+        estimation_step: int = 1,
+        target_update_freq: int = 0,
+        reward_normalization: bool = False,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(model, optim, discount_factor,
+                         estimation_step, target_update_freq,
+                         reward_normalization, **kwargs)
+        self._num_atoms = num_atoms
+        self._v_min = v_min
+        self._v_max = v_max
+        self.support = torch.linspace(self._v_min, self._v_max,
+                                      self._num_atoms)
+        self.delta_z = (v_max - v_min) / (num_atoms - 1)
+
+    @staticmethod
+    def prepare_n_step(
```

Comment:
I'm not quite sure what's the difference between this n_step reimplementation and the existing implementation. Could you please explain more?
Also, a good way is to modify the original implementation instead of a duplicated new one.

Reply:
the existing implementation directly calculates the target_q.  This n_step reimplementation just calculates the n-step return, S_{t+n} and done_{t+n}.

Reply:
In Distributional RL, the target_q is two dimensional array. I think it is inefficient to convert this array many times.

Reply:
So why not extend the original implementation?

Reply:
Since the original implementation is used in most Q-learing based algorithms, I'm cautious about modifying it. Maybe I will change it in next PR after I finish the Atari tests on C51.

Reply:
> I'm cautious about modifying it.

Don't worry about that, I will take care of it.

Reply:
OK. I'll try to modify it later.

Reply:
> I think it is inefficient to convert this array many times.

I think it is negligible.

```python
import time
import numpy as np
from numba import njit

np.random.seed(0)

nstep = 10
rew = np.random.rand(nstep)
target = np.random.rand()
# assume done == False
gamma = .1
bsz = 64
n_atoms = 51

# @njit
def compute_target(gamma, nstep, rew, target):
    for n in range(nstep - 1, -1, -1):
        target = target * gamma + rew[n]
    return target

def timeit(n, func):
    t0 = time.time()
    for _ in range(n):
        func()
    return (time.time() - t0) / n

# single_target
single_target = compute_target(gamma, nstep, rew, target)

# batch_target
batch_target = compute_target(gamma, nstep, rew, np.array([target] * bsz))

# c51_target
c51_target = compute_target(gamma, nstep, rew, np.full([bsz, n_atoms], target))

assert single_target == batch_target[0] == c51_target[0, 0]

# timeit
n = 10000
single_target_time = timeit(n, lambda: compute_target(gamma, nstep, rew, target))
batch_target_time = timeit(n, lambda: compute_target(gamma, nstep, rew, np.array([target] * bsz)))
c51_target_time = timeit(n, lambda: compute_target(gamma, nstep, rew, np.full([bsz, n_atoms], target)))
print(single_target_time)
print(batch_target_time)
print(c51_target_time)
print(c51_target_time / batch_target_time)
```

The result of `c51_target_time / batch_target_time` is about 2~3, however, their order of magnitude is 1e-5, so that I think it is negligible.
I'll modify the current nstep implementation to extend any shape of the given `target_q`.

Reply:
After you modify the current nstep implementation, I can modify the current implementation of C51.